### PR TITLE
fix: hooks: process generated hooks

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [X] Have you successfully run `make lint`?
- [X] Have you successfully run `pytest tests/unit`?

-----
#### Overview
Port support for generated hooks from `snapcraft_legacy` to `snapcraft`.

#### Issue
- Hooks generated during snapcraft's lifecycle were not being included in the snap.
- https://forum.snapcraft.io/t/call-for-testing-snapcraft-7/30201/43

#### Changes
- `core22` snaps now follow same behavior as `core18` and `core20`
  - Hooks generated during the lifecycle are included in the final snap package
  - A wrapper for each generated hook is written to `meta/hooks/`
  - Pre-existing hooks in `snap/hooks/` take priority over generated hooks
- Spread test for generated and pre-existing hooks for `core18`, `core20`, and `core22`

(CRAFT-1159)

### CODE REVIEW

File: snapcraft/parts/setup_assets.py [hunk 1]
```diff
@@ -49,13 +49,16 @@ def setup_assets(
 
     _write_snap_directory(assets_dir=assets_dir, prime_dir=prime_dir, meta_dir=meta_dir)
 
+    # create wrappers for hooks in the snap/hooks directory
+    _create_hook_wrappers(prime_dir)
+
     if project.hooks:
```
Comment (sergiusens) [inline snapcraft/parts/setup_assets.py:55]: Add a code comment here stating that these will replace anything added by _create_hook_wrappers

**Review by sergiusens — COMMENTED** [review]
Looks good, but a few tests that handle conflicts would be nice, a couple scenarios:

- hooks exist in both $PROJECT/snap/hooks and $PRIME/snap/hooks
- hook in $PROJECT/snap/hooks gets updated
- first run has  $PRIME/snap/hooks but second run adds $PROJECT/snap/hooks

this might be nice to add (perhaps separately):

- fast hashing to only update on changes (we might need to eventually track the hook itself and not the destination)
- progress message stating that hook _foo_ is going to be updated because contents changed

File: tests/spread/general/generated-hooks/snap/snapcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+name: hooks-generated
+base: core22
+version: '1.0'
+summary: Test generated hooks.
+description: |
+  Custom hooks should be built, staged in meta/hooks/, and executed.
+  Hooks that exist in snap/hooks/ (and are not pulled or built during snapcraft's lifecycle) take priority over generated hooks.
+  In this project, three hooks are generated - configure, install, and remove.
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply tests/spread/general/generated-hooks/snap/snapcraft.yaml]: Clarified.  I was trying to combine 2 tests into 1 spread test.

**Comment by mr-cal** [comment]
Added 5 spread tests to cover more scenarios.

**Review by sergiusens — APPROVED** [review]
very nice

**Comment by mr-cal** [comment]
I add a commit with minor update to the spread test snap names.

File: snapcraft/parts/setup_assets.py [hunk 3]
```diff
@@ -52,6 +52,7 @@ def setup_assets(
     # create wrappers for hooks in the snap/hooks directory
     _create_hook_wrappers(prime_dir)
 
+    # hooks with command chains will execute the command chain instead of the hook wrapper
```
Comment (sergiusens) [inline snapcraft/parts/setup_assets.py:55]: this comment makes no sense, the hook is still executed

Comment (mr-cal) [reply snapcraft/parts/setup_assets.py:55]: I'd appreciate some clarification on this! I can't find anything that explicitly clarifies command-chains execution for hooks.  I've been looking in the legacy codebase, snapd codebase, and in the documentation.

After testing, I think that the execution of the hook is conditional on the executables in the command-chain ending with `exec "$@"`.  This is the only way it is tested in [snapd](https://github.com/snapcore/snapd/blob/7f63a940dc584fa0ba528986daa2e83a650e0b9e/tests/main/command-chain/command-chain/chain1#L6).

Without `exec $@` at the end of the command-chain scripts, the hook itself doesn't execute.  And there are no warnings or errors.  I can remove the stub hook too, with no errors or warnings.

**Pull request merged**